# WTA Match Predictor

### Contents

- Understanding the Problem Statement
- Data Collection
- Data Checks
- Exploratory Data Analysis
- Preprocessing
- Model Training
- Choose Best Model

### **Understanding the Problem Statement**

- This project seeks to determine what variables affect the outcome of a tennis match on the WTA tour.
- Additionally, this project seeks to train and test a model that can accurately predict the winner of a WTA match.

### **Data Collection**

* The data for this project comes from [Jeff Sackmann of Tennis Abstract.](https://github.com/JeffSackmann/tennis_wta)
* The data is presented such that each observation is a match played, and matches are grouped by year/season. Each observation includes information about each player and their performance in the match.
    - Key statistics include 1st serve percentage, break point conversions, aces/double faults, and more. 

We will be using several libraries to help us explore the data, including ``numpy``, ``pandas``, ``seaborn``, and ``matplotlib``.

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

#### Import Data as Pandas DataFrame

In [6]:
df = pd.read_csv('data/stud.csv')

#### Explore Head and Shape

In [7]:
df.head()

tourney_id  tourney_name surface  draw_size tourney_level  tourney_date  \
0   2023-609  Indian Wells    Hard         96            PM      20230306   
1   2023-609  Indian Wells    Hard         96            PM      20230306   
2   2023-609  Indian Wells    Hard         96            PM      20230306   
3   2023-609  Indian Wells    Hard         96            PM      20230306   
4   2023-609  Indian Wells    Hard         96            PM      20230306   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        286     214544          2.0          NaN  ...    58.0     37.0   
1        285     216347          1.0          NaN  ...    55.0     30.0   
2        284     221054          NaN          NaN  ...    52.0     37.0   
3        283     201514          NaN          NaN  ...    24.0     14.0   
4        282     201614          5.0          NaN  ...    50.0     34.0   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0      16.0    13.0        8.0       11.0            2               6100   
1       4.0    10.0        3.0        8.0            1              10585   
2      10.0    12.0        5.0        8.0           77                784   
3      12.0     8.0        0.0        5.0           83                743   
4      21.0    14.0        1.0        4.0            5               4905   

  loser_rank loser_rank_points  
0         16              2205  
1         36              1303  
2         13              2246  
3         43              1200  
4         49              1080  

[5 rows x 49 columns]

#### Dataset Shape

In [8]:
df.shape

(113, 49)

This CSV stud, which is a sample of the larger dataset, includes 113 observations and 49 variables.

### Dataset Information

- Tourney Information: tourney_id, tourney_name, surface, draw_size, etc.
- Match Information: match_num, winner_id, loser_id, etc.
- Match Statistics: w_1stIn, w_1stWon, w_2ndWon, w_SvGms, w_bpSaved, w_bpFaced, etc.
    - Includes the same statistics for winner/loser

### **Data Checks**
Before working with the data, we must check and address the following:
- Missing values
- Duplicates
- Data types
- Unique values
- Statistics
- Categories

#### Check Missing Values

In [14]:
df.isna().sum()

tourney_id             0
tourney_name           0
surface                0
draw_size              0
tourney_level          0
tourney_date           0
match_num              0
winner_id              0
winner_seed           48
winner_entry          99
winner_name            0
winner_hand            0
winner_ht             30
winner_ioc             0
winner_age             0
loser_id               0
loser_seed            77
loser_entry           81
loser_name             0
loser_hand             0
loser_ht              36
loser_ioc              0
loser_age              0
score                  0
best_of                0
round                  0
minutes                0
w_ace                  1
w_df                   1
w_svpt                 1
w_1stIn                1
w_1stWon               1
w_2ndWon               1
w_SvGms                1
w_bpSaved              1
w_bpFaced              1
l_ace                  1
l_df                   1
l_svpt                 1
l_1stIn                1


#### Check Duplicates

In [13]:
df.duplicated().sum()

np.int64(0)

There are no duplicates values in the data set.

#### Check Data Types

In [16]:
# Check Null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113 entries, 0 to 112
Data columns (total 49 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tourney_id          113 non-null    object 
 1   tourney_name        113 non-null    object 
 2   surface             113 non-null    object 
 3   draw_size           113 non-null    int64  
 4   tourney_level       113 non-null    object 
 5   tourney_date        113 non-null    int64  
 6   match_num           113 non-null    int64  
 7   winner_id           113 non-null    int64  
 8   winner_seed         65 non-null     float64
 9   winner_entry        14 non-null     object 
 10  winner_name         113 non-null    object 
 11  winner_hand         113 non-null    object 
 12  winner_ht           83 non-null     float64
 13  winner_ioc          113 non-null    object 
 14  winner_age          113 non-null    float64
 15  loser_id            113 non-null    int64  
 16  loser_se

#### Check Count of Unique Values

In [18]:
df.nunique()

tourney_id              7
tourney_name            7
surface                 3
draw_size               3
tourney_level           3
tourney_date            5
match_num              38
winner_id              61
winner_seed            18
winner_entry            3
winner_name            61
winner_hand             3
winner_ht              22
winner_ioc             28
winner_age             55
loser_id               96
loser_seed             20
loser_entry             3
loser_name             96
loser_hand              3
loser_ht               22
loser_ioc              36
loser_age              78
score                  77
best_of                 1
round                   6
minutes                70
w_ace                  14
w_df                   14
w_svpt                 62
w_1stIn                48
w_1stWon               35
w_2ndWon               24
w_SvGms                13
w_bpSaved              15
w_bpFaced              18
l_ace                  10
l_df                   15
l_svpt      

### Check Statistics of Dataset

In [19]:
df.describe()

draw_size  tourney_date   match_num      winner_id  winner_seed  \
count  113.000000  1.130000e+02  113.000000     113.000000    65.000000   
mean    50.123894  2.023046e+07  284.185841  210496.292035     6.230769   
std     24.518627  1.139821e+02    9.897284    7052.453367     5.981140   
min     32.000000  2.023031e+07  263.000000  201458.000000     1.000000   
25%     32.000000  2.023040e+07  277.000000  203475.000000     2.000000   
50%     32.000000  2.023040e+07  284.000000  211651.000000     4.000000   
75%     64.000000  2.023061e+07  293.000000  214954.000000     7.000000   
max     96.000000  2.023062e+07  300.000000  232447.000000    29.000000   

        winner_ht  winner_age       loser_id  loser_seed    loser_ht  ...  \
count   83.000000  113.000000     113.000000   36.000000   77.000000  ...   
mean   174.626506   26.423894  210621.433628    9.666667  173.636364  ...   
std      5.547381    4.086110    6756.064422    8.691539    6.290990  ...   
min    160.000000   17.900000  200748.000000    1.000000  160.000000  ...   
25%    170.500000   23.700000  203500.000000    3.750000  170.000000  ...   
50%    175.000000   26.100000  211702.000000    6.500000  175.000000  ...   
75%    178.000000   28.600000  214939.000000   13.500000  178.000000  ...   
max    185.000000   35.600000  232447.000000   32.000000  185.000000  ...   

          l_1stIn    l_1stWon    l_2ndWon     l_SvGms   l_bpSaved   l_bpFaced  \
count  112.000000  112.000000  112.000000  112.000000  112.000000  112.000000   
mean    43.991071   25.866071   10.839286   10.392857    4.642857    9.205357   
std     15.723681   10.907057    5.260003    2.819997    2.995062    3.678862   
min     15.000000    5.000000    0.000000    5.000000    0.000000    3.000000   
25%     31.750000   16.750000    7.000000    8.000000    2.000000    7.000000   
50%     40.500000   25.000000   10.000000   10.000000    4.000000    9.000000   
75%     54.250000   33.250000   14.000000   12.000000    6.000000   11.000000   
max     85.000000   59.000000   26.000000   18.000000   16.000000   24.000000   

       winner_rank  winner_rank_points   loser_rank  loser_rank_points  
count   113.000000          113.000000   113.000000         113.000000  
mean     93.787611         1665.982301   126.194690        1033.424779  
std     138.021120         1720.814072   148.874339        1007.017900  
min       1.000000           26.000000     3.000000          11.000000  
25%      15.000000          563.000000    37.000000         429.000000  
50%      66.000000          851.000000   103.000000         630.000000  
75%     116.000000         2296.000000   154.000000        1258.000000  
max     817.000000        10585.000000  1077.000000        5605.000000  

[8 rows x 35 columns]

### **Exploratory Data Analysis**

asdf

### **Preprocessing**

asdf

### **Model Training**

asdf

### **Choose Best Model**

asdf